src/parser.ts

```diff
@@ -3609,6 +3613,48 @@ export class Parser extends DiagnosticEmitter {
       }
     }
   }
+  parseStringLiteral(tn: Tokenizer, startPos: i32, quote: i32 = -1): Expression {
+    return Node.createStringLiteralExpression(tn.readString(quote), tn.range(startPos, tn.pos));
+  }
+
+  parseTemplateLiteralExpression(tn: Tokenizer): Expression | null {
+    var startPos = tn.pos;
+    tn.inStringTemplate = true;
+
+    // at `(String*${ Expression }*String*)*`
+    const parts: Expression[] = [this.parseStringLiteral(tn, startPos)];
+
+    var token = tn.peek();
+    while (token == Token.DOLLAR) {
+      tn.skip(token);
+      tn.skip(Token.OPENBRACE);
+      let expr = this.parseExpressionStart(tn);
+      if (expr == null) return null;
+      parts.push(expr);
+      tn.skip(Token.CLOSEBRACE);
+      token = tn.next();
+      if (token == Token.TEMPLATELITERAL) {
+        break;
+      }
+      if (token == Token.DOLLAR) {
+        continue;
+      }
+      startPos = tn.pos;
+      parts.push(this.parseStringLiteral(tn, startPos, CharCode.BACKTICK));
+      token = tn.next();
+    }
+    if (token == Token.TEMPLATELITERAL) {
+      tn.advance();
+      tn.inStringTemplate = false;
+    }
+    if (parts.length == 1) {
+      return parts[0];
+    }
+    return parts.reduce((acc: Expression | null, expr: Expression) => {
+      if (acc == null) return expr;
+      return Node.createBinaryExpression(Token.PLUS, acc, expr, tn.range(startPos, tn.pos));
+    }, null);
+  }
```

Comment:
Ideally this would generate a TemplateLiteral node holding the nodes of the template literal plus having a step to compile such a node in the compiler, instead of transpiling. Otherwise the AST cannot be reused for syntax highlighting in a future IDE extension. On such a node, I could imagine properties like `stringParts[N]`, `expressionParts[N-1]`.

Reply:
Ahh good point.  Should we also add a default `toString` to all reference types?

Reply:
My vote is that the expressions must have a type `String` in the `expressionParts` field. Although, it would be nice to check for an empty `toString()` method. I realize this might not be possible.

Reply:
The basic types already have `#toString` implementations, like `I32#toString` for `i32`s, `String#toString` (returning itself) for `string`s etc. Typically one would resolve the expression parts, obtaining their type, and error if there is no `toString` or call it otherwise before concatenating with string parts.

Reply:
I vote for a default toString()  method, that uses something like `nameof<T>()`.

Reply:
It's entirely possible to do something like this:

```ts
if (!isDefined(obj.toString())) ERROR("Must implement toString()");
```

As-pect uses this technique now.

Reply:
In the compiler context, perhaps this might not be so useful, actually.